cookbooks/cdo-repository/libraries/git.rb

```diff
@@ -7,7 +7,7 @@
 # - Better branch names:
 #   - Set checkout branch to remote track the corresponding remote branch with the same name.
 #   - Support switching to a new checkout branch after the initial clone.
-module CDO
+module Cdo
```

Comment:
Just wondering, why do we need a `module Cdo` which only has another `module Provider` nested inside it, as opposed to a `module CdoProvider` or similar?

Reply:
The nested module isn't necessary in this particular case but it's still a convention generally worth consistently following in [Ruby](https://www.oreilly.com/content/ruby-cookbook-modules-and-namespaces/) (and [software namespaces more generally](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/names-of-namespaces)), where the top-level namespace component (`module`) is the name of the company/product to avoid conflicts across dependencies.

From what I've seen, prefixing a `module` with the product name (`CdoProvider`, `Cdo[X]`) is less conventional than nested modules (`Cdo::Provider`, `Cdo::[X]`), possibly in part because of [implicit constant-lookup scope within nested modules](https://docs.ruby-lang.org/en/master/syntax/modules_and_classes_rdoc.html#label-Constants) (e.g., `Cdo::[X]` can reference `Cdo::Provider` just as `Provider`).

While on the subject, it's also worth noting that flattened nesting (e.g., `module Cdo::Provider`) is [generally discouraged](https://github.com/rubocop-hq/ruby-style-guide#namespace-definition) because it requires the top-level module to already be defined, and also has a different constant-lookup scoping (`Cdo::[X]` _cannot_ reference `Cdo::Provider` just as `Provider`).

Reply:
Okay. And if we just had two different `module CDO`, one in our application and one in the cookbook libraries, then that would just end up adding more stuff to the same module, right? Which is actually fine since it satisfies the intention of grouping everything for one company/organization together.

And the problem in this case is that the other `CDO` is a constant variable, not a module, right?

Reply:
Yeah, using the `module` keyword on an existing name is called ['reopening' the module](https://docs.ruby-lang.org/en/2.5.0/syntax/modules_and_classes_rdoc.html#label-Module+Definition).

And yeah the naming problem was that our application uses `CDO.*` (`CDO` is a `Class`, a singleton instance of `Cdo::Impl`) for config, and `Cdo` (`Module`) for namespacing, and you can't reopen a `Class` as a `Module`. These names aren't really ideal (because it's still somewhat confusing that the only difference is capitalization) but that's where our codebase is at today.